cross-chain/bob/contracts/libraries/RateLimiter.sol

```diff
@@ -0,0 +1,217 @@
+    /// @return The token bucket.
+    function _currentTokenBucketState(
+        TokenBucket memory bucket
+    ) internal view returns (TokenBucket memory) {
+        // We update the bucket to reflect the status at the exact time of the
+        // call. This means we might need to refill a part of the bucket based
+        // on the time that has passed since the last update.
+        bucket.tokens = uint128(
+            _calculateRefill(
+                bucket.capacity,
+                bucket.tokens,
+                block.timestamp - bucket.lastUpdated,
+                bucket.rate
+            )
+        );
+        bucket.lastUpdated = uint32(block.timestamp);
+        return bucket;
+    }
+
+    /// @notice Sets the rate limited config.
+    /// @param s_bucket The token bucket
+    /// @param config The new config
+    function _setTokenBucketConfig(
+        TokenBucket storage s_bucket,
+        Config memory config
+    ) internal {
+        // First update the bucket to make sure the proper rate is used for all the time
+        // up until the config change.
+        uint256 timeDiff = block.timestamp - s_bucket.lastUpdated;
+        if (timeDiff != 0) {
+            s_bucket.tokens = uint128(
+                _calculateRefill(
+                    s_bucket.capacity,
+                    s_bucket.tokens,
+                    timeDiff,
+                    s_bucket.rate
+                )
+            );
+
+            s_bucket.lastUpdated = uint32(block.timestamp);
+        }
+
+        s_bucket.tokens = uint128(_min(config.capacity, s_bucket.tokens));
+        s_bucket.isEnabled = config.isEnabled;
+        s_bucket.capacity = config.capacity;
+        s_bucket.rate = config.rate;
+
+        emit ConfigChanged(config);
+    }
+
+    /// @notice Validates the token bucket config
+    function _validateTokenBucketConfig(
+        Config memory config,
+        bool mustBeDisabled
+    ) internal pure {
+        if (config.isEnabled) {
+            if (config.rate >= config.capacity || config.rate == 0) {
+                revert InvalidRateLimitRate(config);
+            }
+            if (mustBeDisabled) {
+                revert RateLimitMustBeDisabled();
+            }
+        } else {
+            if (config.rate != 0 || config.capacity != 0) {
+                revert DisabledNonZeroRateLimit(config);
+            }
+        }
+    }
+
+    /// @notice Calculate refilled tokens
+    /// @param capacity bucket capacity
+    /// @param tokens current bucket tokens
+    /// @param timeDiff block time difference since last refill
+    /// @param rate bucket refill rate
+    /// @return the value of tokens after refill
+    function _calculateRefill(
+        uint256 capacity,
+        uint256 tokens,
+        uint256 timeDiff,
+        uint256 rate
+    ) private pure returns (uint256) {
+        return _min(capacity, tokens + timeDiff * rate);
```

Comment:
Is there a risk of overflow here?

Reply:
The contract uses Solidity ^0.8.4, which includes automatic overflow protection. Any arithmetic overflow will cause the transaction to revert rather than silently wrapping around.